packages/astro/src/core/build/static-build.ts

```diff
@@ -232,18 +276,39 @@ async function clientBuild(
 	return buildResult;
 }
 
-async function cleanSsrOutput(opts: StaticBuildOptions) {
-	const out = getOutDirWithinCwd(opts.settings.config.outDir);
+/**
+ * For each statically prerendered page, replace their SSR file with a noop.
+ * This allows us to run the SSR build only once, but still remove dependencies for statically rendered routes.
+ */
+async function cleanStaticOutput(opts: StaticBuildOptions, internals: BuildInternals) {
+	const allStaticFiles = new Set();
+	for (const pageData of eachPrerenderedPageData(internals)) {
+		allStaticFiles.add(internals.pageToBundleMap.get(pageData.moduleSpecifier));
+	}
+	const ssr = opts.settings.config.output === 'server';
+	const out = ssr ? opts.buildConfig.server : getOutDirWithinCwd(opts.settings.config.outDir);
 	// The SSR output is all .mjs files, the client output is not.
 	const files = await glob('**/*.mjs', {
 		cwd: fileURLToPath(out),
 	});
+
 	if (files.length) {
+		await eslexer.init;
 		// Remove all the SSR generated .mjs files
 		await Promise.all(
 			files.map(async (filename) => {
+				if (!allStaticFiles.has(filename)) {
+					return;
+				}
 				const url = new URL(filename, out);
-				await fs.promises.rm(url);
+				const text = await fs.promises.readFile(url, { encoding: 'utf8' });
+				const [, exports] = eslexer.parse(text);
+				// Replace exports (prerendered pages) with a noop
+				let value = 'const noop = (() => {});';
+				for (const e of exports) {
+					value += `\nexport const ${e.n} = noop;`;
+				}
+				await fs.promises.writeFile(url, value, { encoding: 'utf8' });
```

Comment:
This checks the `prerender` chunk and wipes out everything except the exported pages. This allows us to only run a single SSR build, but eliminate any static-only dependencies from the server before deploy.

This is also why we're writing all static pages to a single chunk. Rollup is smart enough to isolate all of their deps in this chunk as well.